### PR TITLE
created base and target branch columns, refactored service for baseUrl

### PR DESCRIPTION
**Pull request by langdonx**
In app.model.ts, I refactored the url property to baseUri so I could use it for linking directly to the branch (and not just the project). Azure DevOps was returning the branches prefixed with 'refs/heads/', which I have to assume is consistent across most repositories and most uses of Git and is safe to remove via `replace`.

The root of my issues with the previous PR is that your reference statements were prefixed with a space.  When I copied and pasted them into VSCode, it indented the entire file with a space and the diff didn't pick up on it.  I don't believe they need the single space indention, do they?

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ prClient.getPullRequests().then(prs => {
         // Pull Request ID cell
         const tableCellId = document.createElement("td");
         tableCellId.setAttribute("sorttable_customkey", "" + pr.id);
-        tableCellId.innerHTML = "<a href='" + pr.url + "' target='_top'>Pull Request #" + pr.id + "</a>";
+        tableCellId.innerHTML = "<a href='" + (pr.baseUri + pr.projectName + "/_git/" + pr.repo + "/pullRequest/" + pr.id) + "' target='_top'>#" + pr.id + "</a>";
```
Comment (aarondandy) [inline src/app.ts]: This may need some url escaping as `projectName` and `repo` may contain spaces... maybe?

Comment (aarondandy) [reply src/app.ts]: This may apply to the other lines below as well.

File: src/app.ts [hunk 2]
```diff
@@ -32,6 +32,14 @@ prClient.getPullRequests().then(prs => {
         const tableCellRepo = document.createElement("td");
         tableCellRepo.innerText = pr.repo;
         tableRow.appendChild(tableCellRepo);
+        // Base cell
+        const tableCellBaseBranch = document.createElement("td");
+        tableCellBaseBranch.innerHTML = "<a href='" + (pr.baseUri + pr.projectName + "/_git/" + pr.repo + "?version=GB" + pr.baseBranch) + "' target='_top'>#" + pr.baseBranch + "</a>";
```
Comment (aarondandy) [inline src/app.ts]: While this _appears_ to be right, does anybody know what this GB prefix is about? Is there any chance of some users having a different prefix?

Comment (aarondandy) [inline src/app.ts]: A branch name can contain characters such as `/` within it. Probably need to do some URI escaping on there too. I can confirm that in one of our repositories the URL looks like `https://corp.visualstudio.com/ProjectName/_git/Cool%20Business%20App?version=GBbug%2Ffix-the-things`

**Comment by aarondandy** [comment]
I think this is a great feature. We have definitely approved merges between branch `a -> b` assuming it was `a -> master` and it has caused some pain. This will be a great way to have that front and center so the branch name stands out well.

File: src/app.ts (hunk 2 above)
Comment (aarondandy) [reply src/app.ts]: Worst case the new feature does not work for them I suppose. I don't see how it could do harm.

File: src/app.ts (hunk 1 above)
Comment (aarondandy) [reply src/app.ts]: I wonder what other characters people have in their project and repo names... @nshenoy do you have something to test these corner cases (or normal cases :p ) out already?

Comment (aarondandy) [reply src/app.ts]: Given that this is exactly what the previous commit did, if you want to hand wave this away that is fair.

File: src/app.ts (hunk 2 above)
Comment (aarondandy) [reply src/app.ts]: This is the worst URL I could make. I can't even fetch this branch!

```
/Test%20_Project(%5EName.!)/_git/Test%20_Repo(%5EName.!)?version=GB'%2F%3Einjection
```

Comment (aarondandy) [reply src/app.ts]: Assuming the values in `.targetBranch` and `.baseBranch` are the unescaped names, some quick testing with encodeURIComponent shoes that it should do the trick here:

```js
encodeURIComponent("'/>injection") // "'%2F%3Einjection"
encodeURIComponent("Test _Project(^Name.!)") // "Test%20_Project(%5EName.!)"
encodeURIComponent("pb&j") // "pb%26j"
```

**Comment by langdonx** [comment]
Sorry I didn't see these until now.  I'll go through the comments and make the necessary changes.

File: src/app.ts (hunk 1 above)
Comment (langdonx) [reply src/app.ts]: @nshenoy If you check the changes to the service you'll see that `pr.url` is no longer exposed.  Before, it was the URL directly to the pull request.  Now it's the base URL of the Repository.  So in this case `pr.baseUrl` is now used twice (once to link to the PR like before, and once to link to the branches in question).

File: src/app.ts (hunk 2 above)
Comment (langdonx) [reply src/app.ts]: Terrific question.  I'm guessing the "GB" stands for Git branch?  I'm guessing it's an identifier to tell the DevOps back end that the request in question is for a Git repo and not a TFS... uh... Source Control... thing.

Comment (langdonx) [reply src/app.ts]: Good call, I will sprinkle encodeURIComponent where needed.

**Comment by langdonx** [comment]
👍
